### DISCUSSION TITLE
The right way of updating server and clients...

### DISCUSSION THREAD

**Discussion by mygitcrazy** [netbirdio/netbird#762] [Issue Triage]
Question,

What would be the right way of updating the server and the clients?

The reason why I ask this, is because when we last update our selfhosted version from v.014.4 to v.014.5, we first updated some clients and then the server itself.

After the update the ACL rules didn't work anymore, and we had to reinstall every client and reorder the ACL rules to make the clients talk to each other again. 

So my question is, what would be to correct way for the next release to update? First the main server itself and then the clients? I guess....

Thanks by the way for your great work on Netbird, really appreciate it 👍

**Comment by mlsmaycon** [comment]
Hello @mygitcrazy sorry to see that this happened to you. On v0.14.5, we've migrated our custom rules to use [OPA policies](https://www.openpolicyagent.org/docs/latest/), which will provide more flexibility and will allow us to use some advanced features. 

Regarding the upgrade order, they should work in either case. If you upgrade agents or servers, we prepare the migration and agent compatibility to prevent these issues from happening.

Even with our test, we should have noticed something. Please share more information so we can validate a few things and reproduce your case. Can you share the following:
1 - confirm if you are running on v0.14.5 for the agents and the management server.
2 - After upgrading the agents and not the management, were the peers able to communicate according to the ACLs?
3 - example of the access control rules you had in place before the upgrade

**Reply by mygitcrazy** [reply]
Thanks a lot for your quick reply on this, especially because it's sunday today 🙏 

I will make a quick report for you and come back with the answers to your questions within an hour....

**Reply by mlsmaycon** [reply]
Thank you for reporting this over the weekend.

Another information we need is: while facing the issue, did you run netbird status in the clients and the peer list was 0/0 ?

**Reply by mygitcrazy** [reply]
My first discovered findings:

Netbird status on client gives some interesting output, in compare what the management server says.
The version numbered seems to be different on the management server, while client says it updated to v.0.14.5
![netbird status server and client](https://user-images.githubusercontent.com/30736903/227776159-fcf50d7f-13ba-4acc-8794-066878f39bcd.jpg)

I already recovered a lot of clients with new keys and have made new ACL rules, that's why it's hard to go back to the situation where it went wrong yesterday.

I do remember (after updating the management server) i pinged a client, from another client, and it stated back with **"ping: sendmsg: Required key not available"**
At that moment al the ACL rules were untouched, and still the same as before the update went through.

How can i check the management server version? 
Because i rolled back a backup yesterday of the management server to v.0.14.4.
I'm pretty sure that after rolling back, i did the update to v.014.5 again on the management server. 
I want to check the current running version at this moment, just in case....

**Reply by mlsmaycon** [reply]
Hi @mygitcrazy the ping message tell us that was a missing wireguard configuration which is probably related to the ACL issue you've described.

Regarding the version issue, I need to investigate, but can you confirm that you don't have 2 binaries running at the same time with `ps aux | grep netbird`?

The management version can be checked by inspecting the image ID, `docker inspect IMAGE_ID | grep version`

What is mostly odd in the case, is that you should not have to add the peers again, they should continue to be registered. After update, you were able to access the dashboard and list all peers?

**Reply by mygitcrazy** [reply]
Hi @mlsmaycon 
The output of "**ps aux | grep netbird**" had the following outcome:
![output ps aux grep netbird](https://user-images.githubusercontent.com/30736903/227837644-b5f2c3d6-a90b-48a6-96ff-609dc40d8a0c.jpg)

I checked the management server (just in case) and the actual running version as of this moment is on v.0.14.5.

**Your last question was:
After update? you were able to access the dashboard and list all peers?**
After updating the management server, I was able to login and list all peers. Strangely all the peers were showing online, and marked green. But on the other hand, clients were reporting that they couldn't connect to each other. So i'm sure it has something to do with the new ACL settings, that came after the update. One thing noticed, is that after reinstalling the client (with new key) and replacing the old ACL rules for new ones, things got working again......

**Reply by mlsmaycon** [reply]
Thanks @mygitcrazy. Regarding the rules, could you share some example of the ACL rules in place?

**Reply by mygitcrazy** [reply]
Sure, I'm glad i can assist and give something back for using Netbird 😉

odd question maybe, but how do you want me to specify the ACL rules that i have in place? 
Something like:

<html>
<body>


Name | Status | Sources | Direction | Destinations |  
-- | -- | -- | -- | -- | --
  |   |   |   |   |  
Allow connection between Proxyservers1+2 | enabled | proxy-server1 | <--direction--> | proxy-server2 |  
disallow connection between all peers | disabled | All | <--direction-->  | All |  


</body>
</html>


**Now proxyserver1 and proxyserver2 are assigned to two groups;**
proxy-server1 is assigned to Group:ALL and proxy-server1
proxy-server2 is assigned to Group:ALL and proxy-server2


Let me know if you need more info on this...

**Reply by mlsmaycon** [reply]
That description looks good. Do you have more ACLs?

**Reply by mygitcrazy** [reply]
I do have more ACL rules...

<html>
<body>


Name | Status | Sources | Direction | Destinations |  
-- | -- | -- | -- | -- | --
  |   |   |   |   |  
Allow connection between Proxyservers1+2 | enabled | proxy-server1 | <--direction--> | proxy-server2 |  
Allow connection between Proxyservers1+3 | enabled | proxy-server1 | <--direction--> | proxy-server3 |
Allow connection between Proxyservers2+3 | enabled | proxy-server2| <--direction--> | proxy-server3 | 
Allow connection between Proxyservers3+4 | enabled | proxy-server3 | <--direction--> | proxy-server4 |   
disallow connection between all peers | disabled | All | <--direction-->  | All |  


</body>
</html>


**Now all these proxyservers are also assigned into two groups;**
proxy-server1 is assigned to Group:ALL and proxy-server1
proxy-server2 is assigned to Group:ALL and proxy-server2
proxy-server3 is assigned to Group:ALL and proxy-server3
proxy-server4 is assigned to Group:ALL and proxy-server4

I do have more rules that are comparable to this....

**Reply by mlsmaycon** [reply]
Thank you @mygitcrazy we will reproduce and validate these cases

**Reply by mygitcrazy** [reply]
It's been a pleasure to help you.
If you need me for something else, just give me a sign and I'll be there 😉

**Reply by mlsmaycon** [reply]
@mygitcrazy we've released a fix for the issues you faced. The latest v0.14.6 contains the fix, please upgrade you management server.

**Reply by mygitcrazy** [reply]
@mlsmaycon hi there, 
Unfortunately I was down with the flu this week, so wasn't able to test the new version. Once I feel a little better, I come back with the update results...

**Reply by mlsmaycon** [reply]
no worries @mygitcrazy , get well soon

**Reply by mygitcrazy** [reply]
@mlsmaycon I have updated every client now, included the main interface and I can report that everything is fully functional.
I will keep an eye on everything, but for now all seems OK.

Thanks for your great support😉👍

Ps
I will close this discussion in a few days...

**Reply by mlsmaycon** [reply]
That's great @mygitcrazy, thanks for checking and providing the information needed to fix the issue.
